torch/_functorch/aot_autograd.py

```diff
@@ -1077,6 +1078,127 @@ def forward(*runtime_args: tuple[Any]):
+    traditional torch.compile joint graph pass for two reasons:
+
+        1. The sharding of parameters must be decided before parameter
+           initialization / checkpoint load, far before torch.compile would
+           ordinarily run.
+
+        2. We need to change the meaning of parameters (e.g., we might replace
+           a replicated parameter with a sharded version of it, changing its
+           input size).  torch.compile is ordinarily semantics preserving, and
+           not allowed to change the meaning of inputs.
+
+    Some descriptors can be quite exotic, so we recommend thinking carefully
+    if there is a safe fallback you can apply to descriptors you don't understand.
+    For example, you should have some way to handle not finding a particular
+    input exactly as is in the final FX graph inputs.
+
+    Note: When using this API, you must create and enter an ExitStack context
+    manager, which will be passed into this function.  This context manager
+    must remain active if you .  (TODO: We may relax this requirement by
+    having AOTAutograd keep track of how to reconstruct all the context
+    managers at a later point in time.)
+
+    NB: You're not obligated to do a /full/ compile in stage2; instead you can
+    leave the forward/backward compilers unspecified in which case the
+    partitioned FX graphs will directly run.  The overall autograd Function
+    can be allowed in graph so you can reprocess it in the context of a
+    (potentially larger) compiled region later.
+
+    NB: These APIs do NOT hit cache, as we only ever cache the final compile results,
+    not the intermediate export result.
+    """
+
+    (
+        functional_call,
+        params_flat,
+        fake_flat_args,
+        full_args_descs,
+        aot_config,
+        fake_mode,
+        shape_env,
+    ) = prepare_aot_module_simplified(
+        mod,
+        args,
+        None,
+        None,
+        None,
+        decompositions,
+        keep_inference_input_mutations,
+        None,
+        None,
+        ignore_shape_env,
+    )
+
+    # TODO: Maybe this should be in create_aot_state?  Not sure, that would
+    # increase its scope
+    stack.enter_context(compiled_autograd._disable())
+
+    aot_state = create_aot_state(
+        stack,
+        functional_call,
+        fake_flat_args,
+        full_args_descs,
+        aot_config,
+        fake_mode,
+        shape_env,
+    )
+    # NB: no cache lookup!
+    aot_graph_capture = aot_stage1_graph_capture(aot_state, functional_call)
+
+    # Invariant: the compiled function takes all of params_flat as argument,
+    # you cannot change structure of this but you can change the inputs
+
+    return aot_state, aot_graph_capture
+
+
+def aot_compile_joint_with_descriptors(
+    aot_state: AOTState, aot_graph_capture: AOTGraphCapture
+) -> callable:
+    """
+    Companion function for aot_export_joint_with_descriptors which compiles the joint
```

Comment:
i'm not familiar with AOTGraphCapture type.  Is the point of separating export_joint from compile_joint functions into 2 stages to allow some user intervention inbetween? or is this just an implementation detail?